### PR TITLE
✨ refactor: cluster scope and service refactor for future EKS support

### LINKED ISSUE #1796: Refactor clusterscope/ec2 service
/kind feature

**Describe the solution you'd like**
Refactor the ClusterScope and ec2 service so that the networking specific side of things can be used in multiple places (i.e. AWSCluster controller and a future AWSManaged Cluster). I'm thinking that we introduce smaller targetted scopes that can be combined if required. For example a ClusterScope has a NetworkScope, and also a ManagedClusterScope has a networkscope.

This would help with #1787 

**Anything else you would like to add:**
The Azure provider has started to refactor their scopes/services into small units.

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by richardcase**
**What this PR does / why we need it**:
A slight refactor of the `ClusterScope` related stuff so that the ec2 service can be reused by the EKS implementation. Specifically in the future there will be a `AWSManagedCluster\AWSManagedControlPlane`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1796

### CODE REVIEW

File: pkg/cloud/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	ListOptionsLabelSelector() client.ListOption
+	PatchObject() error
+	Close() error
+	APIServerPort() int32
+	AdditionalTags() infrav1.Tags
+	SetFailureDomain(id string, spec clusterv1.FailureDomainSpec)
+	Bastion() *infrav1.Bastion
+	SetBastionInstance(instance *infrav1.Instance)
+	SSHKeyName() *string
+}
```
Comment (vincepri) [inline pkg/cloud/interfaces.go:67]: Might be worthwhile to add comments to each of these, given that the interface will be published in godoc

File: pkg/cloud/scope/clients.go [hunk 2]
```diff
@@ -17,16 +17,74 @@ limitations under the License.
...
+	resourceTagging.Handlers.Build.PushFrontNamed(getUserAgentHandler())
+	resourceTagging.Handlers.Complete.PushBack(recordAWSPermissionsIssue(target))
+
+	return resourceTagging
+}
+
+// NewSecretsManagerClient creates a new Secrets API client for a given session
+func NewSecretsManagerClient(session cloud.Session, target runtime.Object) secretsmanageriface.SecretsManagerAPI {
+	sClient := secretsmanager.New(session.Session())
+	sClient.Handlers.Build.PushFrontNamed(getUserAgentHandler())
```
Comment (randomvariable) [inline pkg/cloud/scope/clients.go:68]: Review the rebasing again, as it looks like the metrics collection from https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1807 has got wiped out here.

File: pkg/cloud/interfaces.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+type Session interface {
+	Session() awsclient.ConfigProvider
+}
+
+// ClusterObject represents a AWS cluster object
+type ClusterObject interface {
+	conditions.Setter
+}
+
+// ClusterScoper is the interface for a cluster scopde
```
Comment (rudoi) [inline pkg/cloud/interfaces.go:41]:
```suggestion
// ClusterScoper is the interface for a cluster scope
```

File: pkg/cloud/interfaces.go (hunk 1 above)
Comment (richardcase) [reply pkg/cloud/interfaces.go]: I've added comments to the interface.

File: pkg/cloud/scope/clients.go (hunk 2 above)
Comment (richardcase) [reply pkg/cloud/scope/clients.go]: @randomvariable - ah yes, dodgey rebase by me. Sorry about that, metrics re-added to client creation.

File: pkg/cloud/interfaces.go (hunk 3 above)
Comment (richardcase) [reply pkg/cloud/interfaces.go]: Corrected the typo.

**Comment by rudoi** [comment]
FWIW built this into an image and rolled it out to an existing self-managed cluster. Tested a handful of Machine/AWSMachine operations and saw no issues!

File: pkg/cloud/interfaces.go [hunk 4]
```diff
@@ -33,35 +33,62 @@ type Session interface {
...
 	Bastion() *infrav1.Bastion
+	// SetBastionInstance sets the bastion instance in the status of the cluster.
 	SetBastionInstance(instance *infrav1.Instance)
+	// SSHKeyName returns the SSH key name to use for instances.
 	SSHKeyName() *string
+
+	// PatchObject persists the cluster configuration and status.
+	PatchObject() error
+	// Close closes the current scope persisting the cluster configuration and status.
+	Close() error
```
[the remark this answers is not in the document]
Comment (richardcase) [reply pkg/cloud/interfaces.go]: I think we should split the ec2 service in the future which would result in this interface being divided up. I didn't do that with this PR as I kept the change to a minimum to support the what was needed for the EKS implementation.

I could do it as part of this or create a new issue?

File: pkg/cloud/interfaces.go [hunk 5]
```diff
@@ -33,35 +33,62 @@ type Session interface {
 	Session() awsclient.ConfigProvider
 }
 
+// ScopeUsage is used to indicate which controller is using a scope
+type ScopeUsage interface {
+	// ControllerName returns the name of the controller that created the scope
+	ControllerName() string
+}
```
[the remark this answers is not in the document]
Comment (richardcase) [reply pkg/cloud/interfaces.go:40]: I originally didn't have this split out. In the end I thought it best to split it out as 'ControllerName' didn't feel like it lived with the other members of the ClusterScoper interface as it's not related to the actual cluster.

File: pkg/cloud/interfaces.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (richardcase) [reply pkg/cloud/interfaces.go]: @randomvariable - is it worth splitting up the ec2 service now or wait until the new AWS services operator is out?

[the remark this answers is not in the document]
Comment (randomvariable) [reply pkg/cloud/interfaces.go]: Unsure tbh, the flip side of the earlier conversation being we don't have clear timelines on the AWS Controllers for Kubernetes project (now tracking in #1818) so can't slot it into the roadmap at present.

Two options:
Leave as the admittedly large interface now:
Minimises changes to current controllers, but future refactoring also affects EKS controllers

Refactor now:
Larger change to current controllers, but may get thrown away by anyway by combination of #1818 and #1484

[the remark this answers is not in the document]
Comment (richardcase) [reply pkg/cloud/interfaces.go]: How about as a middle ground i split out the network specific stuff from the ec2 service into a new service and change the interface above to move the network related stuff into its own interface?

[the remark this answers is not in the document]
Comment (richardcase) [reply pkg/cloud/interfaces.go]: @randomvariable @vincepri - i've trimmed down this interface and split out the network parts of the ec2 service into their owm service (with scope).

**Pull request merged**